### ISSUE TITLE
tmpfs (/tmp) FS problems with QEMU - temporary dir location to be changed

### ISSUE THREAD

**Issue by svirt** [avocado-framework/avocado#895]
HI **@lmr, @ldoktor**

Recently I faced a case (running test virsh.blockpull.normal_test.network_disk.iscsi.notimeout.nobase.async , other similar tests maybe impacted ) where it was failing with **"TestFail: error: internal error: unable to execute QEMU command 'transaction': Could not open '/tmp/avocado_YM43eu/blk_src_file.snap1': Invalid argument"** . After long debugging (as the QEMU message is not much clear) I found out that the problem is related to the case when iscsi disk is added during the test and it has the argument **cache='none'** , which is causing problems with iscsi disk-file which is located in /tmp/avocado_XXXX (according to the new tmp dir approach from data_dir.py, instead of using the tmp dir in home dir). /tmp is type of tmpfs which doesn't support O_DIRECT, according to the log below
tail avocado-vt-vm1.log
...
...
2015-11-27T09:43:59.957100Z file system may not support O_DIRECT
qemu: terminating on signal 15 from pid 26243
2015-11-27 09:44:00.567+0000: shutting down

Thus qemu is not able to manipulate properly with files located in /tmp (tmpfs). Maybe it's better to move back tmp dir to a home dir ?

**Comment by ldoktor** [comment]
Hmm, yes, this is actually a problem for many tests. Actually I'd have nothing against using `data_dir.get_base_dir() + "tmp"`. We already use the base dir for other things...

**Comment by ldoktor** [comment]
^^ btw a workaround for you is to set an environment variable `TMP=/var/tmp` (or other location) and avocado will start using this location instead.

**Comment by svirt** [comment]
Thanks for your notes, yes, **data_dir.get_base_dir() + "tmp"** should be suitable for this case...

**Comment by ldoktor** [comment]
@lmr, @clebergnu any objections?

**Comment by svirt** [comment]
I have here a small note/concern, but it's a probably particular case, the tmp dir requires around ~11 GB for some tests (for iscsi disk-files e.g.) and the / partition should be pretty big in order to have enough space in /usr/share/avocado/.... dir.

**Comment by lmr** [comment]
@ldoktor @clebergnu  I told you that tmp on tmpfs causes test problems, that's why the code had the provision of creating the avocado tmp dir on a non tmpfs filesystem :)

I have an idea to solve this - what about we revert the patches for custom tmp handling?

**Comment by lmr** [comment]
So I guess my suggestion is in the "I have an objection" category :)

**Comment by lmr** [comment]
@svirt, we can solve this easily 2 ways:

1) In your setup - You mask tmp on tmpfs -> `systemctl mask tmp.mount` in your test systems
2) In avocado's code - reverting some commits that changed this recently, see previous comments.

I have objections to tmp dir in data_dir, mainly, data dir is for more permanent user data, and it's common that people simply mount that on an NFS share or whatnot. `/var/tmp` is a much better location for that.

**Comment by ldoktor** [comment]
Well I definitely disagree with 1 (better solution for me would be to `export TMP=/var/tmp`. About 2, yes, probably we might revert it (or use some cleaner way with a proper comment to avoid yet another round of remove+revert...)

**Comment by ldoktor** [comment]
PS: In the world where I'm the only avocado user/developer, I'd have chosen the `export TMP=/var/tmp` approach...

**Comment by clebergnu** [comment]
IMHO it's a mistake to make a framework design decision (that will affect
most users and tests) based on a number of (exceptional) use cases.

This is similar to the `public` parameter added to tmp directory creation
while ago in avocado-vt only: it should not interfere with the **avocado**
default APIs.

**Comment by ldoktor** [comment]
well not really. In testing you usually rely on temp dirs, put the stuff under testing there. It will be affected if one uses in-memory-tmpfs. That's why I said we might put it into `avocad_datadir`. Tak our doublefree examples, we compile the sources and put the result into tmpdir, testing in-memory object. It will definitely behave differently than file stored on a rotary disk.

So for me the solutions are (accordingly to my preferences):

1. leave it up to users to set `TMP` (documentation)
2. use avocado `basedir` (~/avocado/tmp/)
3. implement different tmpdir for avocado-vt

**Comment by lmr** [comment]
I don't think there cases are so exceptional. The decision the systemd team took may have its justifications, but it doesn't change the fact that for years /tmp was a directory mounted on a disk-backed filesystem. People got used to it, and that has been changed recently, forcing adaptation. I like most of the design decisions the systemd team does, so I won't complain endlessly about the few things I don't like.

I don't like putting tmp stuff in the datadir, because datadir was meant to store more durable artifacts used by the tests, anything generated *by* the tests shouldn't be there (allowing it to be basically read only). Of course I am just one person, and your votes can override mine.

In the end, we can just do nothing, and eventually, as a public service, tell people that they can do `systemctl mask tmp.mount` in their test systems, should the need arise, or suggest users to set `TMP` to accomplish a similar goal. It's all fine by me.

The solution of implementing different tmpdir for avocado-vt is not a good one, as many other situations people might use `O_DIRECT` in files saved to a tmp location. It's not specific to virt testing, and it comes up more often than you give it credit for.

**Comment by ldoktor** [comment]
OK and how about this:

```patch
diff --git a/scripts/avocado b/scripts/avocado
index 4493889..3d0d21d 100755
--- a/scripts/avocado
+++ b/scripts/avocado
@@ -81,5 +81,15 @@ if os.path.isdir(os.path.join(basedir, 'avocado')):
 from avocado.core.app import AvocadoApp
 
 if __name__ == '__main__':
+    # Override tmp in case it's not hardcoded
+    for attr in ("TMP", "TEMP", "TMPDIR"):
+        if attr in os.environ:
+            break
+    else:   # TMP not set by user, use /var/tmp if exists
+        # TMP not set by user in environment. Try to use /var/tmp to avoid
+        # possible problems with "/tmp" being mounted as TMPFS without the
+        # support for O_DIRECT
+        if os.path.exists("/var/tmp"):
+            os.environ["TMP"] = "/var/tmp"
     app = AvocadoApp()
     sys.exit(app.run())
```

It'd only be used with our runner and it would not override the value set by user.

**Comment by svirt** [comment]
**@ldoktor**
Hi, my personal opinion is that it seems good...

**Comment by lmr** [comment]
@ldoktor Oh well, fine by me, seems reasonable. When you have time, please send it as a patch so that we can resolve this issue once for all.

**Comment by lmr** [comment]
Solved by merging #911, I'm closing this issue.
